### PR TITLE
[MNT] Docker files updated

### LINKED ISSUE #5359: [ENH] update base docker images
We should consider updating the base images of all docker files. The current base image is bullseye and I suggest upgrading it to bookworm, which is the current stable Debian version.

@yarnabrina has also made two excellent suggestions, which I'm currently in favor of:
1. Not mentioning the patch versions to ensure that we always get the latest one.
2. Using slim versions of the base image to reduce the overall size of the created image. We might have to manually install a few required packages but once we have the necessary packages, it should take minimal effort to maintain the dockerfiles.

Overall, I consider this as an easy and quick win.

Edit: I missed one of the points that @yarnabrina mentioned [here](https://github.com/sktime/sktime/pull/5356#issuecomment-1746594692), it was to consider using `all_extras_pandas2` instead of `all_extras`. I don't have an opinion on this yet but we should also discuss it in this issue.

### PR DESCRIPTION
**Pull request by deysanjeeb**
#### Reference Issues/PRs
Fixes #5359. 

#### What does this implement/fix? Explain your changes.
Updated the dockerfiles to use bookworm-slim images.

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

**Review by yarnabrina — CHANGES_REQUESTED** [review]
Thanks for your contribution.

1. Please do not use such a specific image, as then we need to create such issues again and again in future. Currently your changes use a specific patch version on a specific Debian version. If you change to a general format, e.g. from `python:3.11.8-slim-bookworm` to `python:3.11-slim`, it'll now use latest patch release of python 3.11 on latest Devian version. 

2. It's not recommended to use `apt` in non-interactive cases, as its functionality is not as stable as that of `apt-get`. Please use that instead.

3. To keep the image sizes minimal, you can consider using `--no-install-recommends` during installation and clean up the cache afterwards.

4. (Non-blocking) Can you list the different package names in alphabetical order please?

5. Please use `all_extras_pandas2` instead of `all_extras`, as `sktime` now uses pandas v2 by default.

**Comment by yarnabrina** [comment]
FYI @achieveordie for more optimisations.

Also, how do we test this? In theory, what should we need to check, I'm confused even in that part. Plus, how do we do that in CI?

File: build_tools/docker/py311.dockerfile [hunk 1]
```diff
@@ -1,8 +1,9 @@
-FROM python:3.11.2-bullseye
+FROM python:3.11.8-slim-bookworm
```
Comment (achieveordie) [inline build_tools/docker/py311.dockerfile:1]: It's not a good idea to mention the patch version, as mentioned by @yarnabrina. Simply using `python:3.11-slim-bookworm` will suffice.

File: build_tools/docker/py38.dockerfile [hunk 2]
```diff
@@ -1,8 +1,9 @@
-FROM python:3.8.16-bullseye
+FROM python:3.8.18-slim-bookworm
```
Comment (achieveordie) [inline build_tools/docker/py38.dockerfile:1]: Likewise, we might have another security patch before its EOL (later this year), so it'll be a good idea to exclude patch version 18 from the docker image.

File: build_tools/docker/py311.dockerfile [hunk 3]
```diff
@@ -1,8 +1,9 @@
...
 
 WORKDIR /usr/src/sktime
 
 COPY . .
 
+RUN apt update && apt install -y gcc build-essential git
 RUN python -m pip install -U pip
 RUN python -m pip install .[all_extras,dev,binder]
```
Comment (achieveordie) [inline build_tools/docker/py311.dockerfile:9]: It should be `all_extras_pandas2` as mentioned by @yarnabrina

**Review by achieveordie — CHANGES_REQUESTED** [review]
Thanks for the PR @deysanjeeb! You missed addressing some parts mentioned in the initial review by @yarnabrina. I've pointed it out for your help.

**Review by achieveordie — COMMENTED** [review]
Perhaps, it will also be worthwhile to add a Dockerfile for Python 3.12 since we officially support that as well.

**Comment by achieveordie** [comment]
> Also, how do we test this? In theory, what should we need to check, I'm confused even in that part. Plus, how do we do that in CI?

That's a very valid question and I'm afraid I don't have an answer for it at the moment. Could you remind me of the original purpose of having these dockerfiles in the first place? I think we had a similar conversation but I could not find it. 

If we can answer what purpose it serves, we might be able to better find out how to test it in CI.

**Comment by yarnabrina** [comment]
> If we can answer what purpose it serves, we might be able to better find out how to test it in CI.

I don't know, I thought you might. Seems I asked earlier as well (https://github.com/sktime/sktime/pull/5426#issuecomment-1764714439) but probably got lost in the discussion. 

@fkiraly can you please help? I just found this, but don't know if anyone actually uses this or not given the amount of time it must take to run full test suite locally.

https://www.sktime.net/en/v0.27.0/developer_guide/continuous_integration.html#alternative-dockerized-testing

**Review by deysanjeeb — COMMENTED** [review]
Hi I made the requested changes to py3.8, py3.9, py3.10, py3.11.

**Comment by fkiraly** [comment]
> Could you remind me of the original purpose of having these dockerfiles in the first place? I think we had a similar conversation but I could not find it.

I suppose, running the tests, or developing `sktime` in your cloud.
Aalthough, personally, I'd just write my own `pip install` script in that case - I mean, these dockers don't really do much.

Originally, multiple people kept adding them, incl some frequent developers, so I did not mind it.

Here are the historical PR:
https://github.com/sktime/sktime/pull/2461
https://github.com/sktime/sktime/pull/2249

Unfortunately, these don't seem to contain a statement of purpose of usage, a hint might be this paragraph in the instructions for testing:
https://www.sktime.net/en/stable/developer_guide/continuous_integration.html#alternative-dockerized-testing

Originally, the contributor of #2249 wanted to remove instrutions for local tesing and completely replace them with instructions for testing via the dockers, which I found a bit extreme (since most contributors afaik test locally), so I moved the contribution to the paragraph "alternatives".

**Comment by fkiraly** [comment]
> I just found this, but don't know if anyone actually uses this or not

Well, there's Hyrum's law
https://www.hyrumslaw.com/
which states that, for a piece of software that is used widely enough, every observable behaviour will be used by someone.

So, worst case, someone out there somewhere has an automatic workflow that gets these docker images from the repo clone and fires them up, assuming they are well-maintained. When we remove them, some real life factory goes up in flames.

Obviously I'm just joking, what I really want to say is that it's not implausible that someone uses these in a reasonable way.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Looks good to me!

may I kindly ask to also update the table in
`docs/source/developer_guide/continuous_integration.rst`?

You can also claim the `maintenance` badge!

**Comment by deysanjeeb** [comment]
Updated the table in ```docs/source/developer_guide/continuous_integration.rst```

**Review by fkiraly — APPROVED** [review]
Thanks!

**Pull request merged**